core/ir/synthetic/decode.c

```diff
@@ -0,0 +1,138 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* decode.c -- a decoder for DR synthetic IR */
+
+#include "decode.h"
+
+#include "../globals.h"
+#include "encode_api.h"
+#include "encoding_common.h"
+#include "instr_api.h"
+#include "opnd_api.h"
+
+/* Decodes the raw bytes of an encoded instruction \p encoded_instr into DR instruction
+ * representation \p instr.
+ * Returns the next instruction's PC.
+ * The encoding scheme followed is described in core/ir/synthetic/encoding_common.h.
+ */
+byte *
+decode_from_synth(dcontext_t *dcontext, byte *encoded_instr, instr_t *instr)
+{
+    /* Interpret the first 4 bytes of encoded_instr (which are always present) as a uint
+     * for easier retrieving of category, eflags, #src, and #dst values.
+     * We can do this safely because encoded_instr is 4 bytes aligned.
+     */
+    uint encoding_header = *((uint *)&encoded_instr[0]);
+
+    /* Decode number of register destination operands.
+     */
+    uint num_dsts = encoding_header & DST_OPND_MASK;
+
+    /* Decode number of register source operands.
+     */
+    uint num_srcs = (encoding_header & SRC_OPND_MASK) >> SRC_OPND_SHIFT;
+
+    instr_set_num_opnds(dcontext, instr, num_dsts, num_srcs);
+
+    /* Decode arithmetic flags.
+     */
+    uint eflags = (encoding_header & FLAGS_MASK) >> FLAGS_SHIFT;
+    uint eflags_instr = 0;
+    if (TESTANY(SYNTHETIC_INSTR_WRITES_ARITH, eflags))
+        eflags_instr |= EFLAGS_WRITE_ARITH;
+    if (TESTANY(SYNTHETIC_INSTR_READS_ARITH, eflags))
+        eflags_instr |= EFLAGS_READ_ARITH;
+    instr->eflags = eflags_instr;
+
+    /* Declare the eflags to be valid.
+     * This is needed in order to retrieve their value without trying to compute it again.
+     */
+    instr_set_arith_flags_valid(instr, true);
+
+    /* Decode synthetic opcode as instruction category.
+     */
+    uint category = (encoding_header & CATEGORY_MASK) >> CATEGORY_SHIFT;
+    instr_set_category(instr, category);
+
+    /* Decode register operand size, if there are any operands.
+     */
+    uint num_opnds = num_dsts + num_srcs;
+    opnd_size_t max_reg_size = OPSZ_NA;
+    if (num_opnds > 0)
+        max_reg_size = (opnd_size_t)encoded_instr[OP_SIZE_INDEX];
+
+    /* Decode register destination operands, if present.
+     */
+    for (uint i = 0; i < num_dsts; ++i) {
+        reg_id_t dst = (reg_id_t)encoded_instr[i + OPND_INDEX];
+        opnd_t dst_opnd = opnd_create_reg((reg_id_t)dst);
+        opnd_set_size(&dst_opnd, max_reg_size);
```

Comment:
We don't need this, right?  Esp when it may not match the reg size.  Ditto below.

Reply:
Right, not needed anymore. Done.